t_stress/test_slow_read.py

```diff
@@ -0,0 +1,226 @@
+
+                        for i in range(1, 190, 2):
+                            c.send_headers(i, headers, end_stream=True)
+                            s.sendall(c.data_to_send())
+
+                        response_stream_ended = False
+
+                        while not response_stream_ended:
+                            # read raw data from the socket
+                            data = s.recv(65536 * 1024)
+                            if not data:
+                                break
+
+                            # feed raw data into h2, and process resulting events
+                            events = c.receive_data(data)
+                            for event in events:
+                                if isinstance(event, h2.events.DataReceived):
+                                    # update flow control so the server doesn't starve us
+                                    c.acknowledge_received_data(
+                                        event.flow_controlled_length, event.stream_id
+                                    )
+                                    win = WindowUpdateFrame(event.stream_id, 1).serialize()
+                                    s.sendall(win)
+                                    time.sleep(0.003)
+                                if isinstance(event, h2.events.StreamEnded):
+                                    # response body completed, let's exit the loop
+                                    response_stream_ended = True
+                                    break
+                            # send any pending data to the server
+                            s.sendall(c.data_to_send())
+
+        elif name == "tcp":
+
+            def run_test():
+                context = ssl.create_default_context()
+                context.set_alpn_protocols(["h2"])
+                context.check_hostname = False
+                context.verify_mode = ssl.CERT_NONE
+
+                with socket.create_connection((hostname, port)) as sock:
+                    sock.setsockopt(socket.SOL_SOCKET, socket.SO_RCVBUF, 16)
+                    with context.wrap_socket(sock, server_hostname="tempesta-tech.com") as s:
+                        c = h2.connection.H2Connection()
+                        settings = dict()
+                        settings[SettingCodes.INITIAL_WINDOW_SIZE] = (1 << 31) - 10
+                        c.local_settings = Settings(initial_values=settings)
+                        c.local_settings.update(settings)
+                        c.initiate_connection()
+                        s.sendall(c.data_to_send())
+
+                        for i in range(1, 190, 2):
+                            hf = HeadersFrame(
+                                stream_id=i,
+                                data=headers_data,
+                                flags=["END_HEADERS", "END_STREAM"],
+                            ).serialize()
+                            s.sendall(hf)
+
+                        time.sleep(63)
```

Comment:
```suggestion
            client = self.get_client("deproxy")
            client.update_initial_settings()
            client.send_bytes(client.h2_connection.data_to_send())
            self.assertTrue(client.wait_for_ack_settings())
            
            client.readable = lambda: False # disable reading from socket
            for _ in range(10):
                client.make_request(
                    client.create_request(
                        method='GET', headers=[], authority="tempesta-tech.com",
                        uri="/large.file"
                    ),
                    end_stream=True,
                )
            
            self.assertFalse(client.wait_for_response(3)) # keepalive_timeout in TempestaFW  config is 2
            client.readable = lambda: True # enable reading from socket
            self.assertTrue(client.wait_for_connection_close())
```

But I think this case checks `keepalive_timeout` directive. And you should add checks for CPU and memory using.

Reply:
Both window-size-based and tcp-based slow-read have the same effect.

Moreover, there are two workarounds to slove slow-read, which also have the same effect:
1. small `keepalive_timeout`
2. small `http_body_len`

Reply:
That's why I suggested creating several http2 streams with a large response body and not reading data from the socket `client.readable = lambda: False`. But we should understand that TempestaFW will not close the connection until the client sends any data.

Reply:
> But we should understand that TempestaFW will not close the connection until the client sends any data.

I don't get your point. If the client does not send any request data, the connection will be closed after the keepalive timer expires, before that, the connection is not harmful, i.e. it does not cause OOM. If the client already sent out the request, then the slow read can be controlled by the `http_body_len` very well.

Reply:
Sorry, I was wrong. I thought `keepalive_timeout` work as frang directives and it has not timer, but it is not.